### ISSUE TITLE
[CURATOR-469] Give background task a chance to execute in CuratorFramework.close()

### ISSUE THREAD

**Issue by jira-importer** [apache/curator#989]
<p>The current impl of CuratorFramework.close() does not really give background task a chance to finish.  The following steps are currently taken in close().</p>

<p> 1. Set the state to STOPPED<br/>
 2. Call executorService.shutdownNow();<br/>
 3. Call executorService.awaitTermination(maxCloseWaitMs, TimeUnit.MILLISECONDS);</p>

<p>After step 1 is complete any background task accessing curator will get an IllegalStateException (See <a href="https://issues.apache.org/jira/browse/CURATOR-467" title="Avoid logging errors after curator framework is closed." class="issue-link" data-issue-key="CURATOR-467">CURATOR-467</a>).  Step 2 interrupts actively running task and dequeues any task waiting run.  In step 3 I wonder why bother to wait?</p>

<p>Making close do the following is one possible way to give background task a chance to run.</p>

<p> 1. Call executorService.shutdown();<br/>
 2. Call executorService.awaitTermination(maxCloseWaitMs, TimeUnit.MILLISECONDS);<br/>
 3. Set the state to STOPPED<br/>
 4. Call executorService.shutdownNow();</p>

<p>Step 1 prevents new task from being added, but gives currently running and queued task a chance.  In step 2 we wait up to the user configured time for task to complete.  In step 3 and 4 we cause any background task that are still running to fail.</p>

---
<details><summary><i>Originally reported by <a title="Keith Turner" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kturner">kturner</a>, imported from: <a href="https://issues.apache.org/jira/browse/CURATOR-469" target="_blank">Give background task a chance to execute in CuratorFramework.close()</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jordan Zimmerman" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=randgalt">randgalt</a>
<li><b>status</b>: Open
<li><b>priority</b>: Major
<li><b>resolution</b>: Unresolved
<li><b>imported</b>: 2025-01-21
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=cammckenzie">cammckenzie</a>:</i>
<p>This sounds like a sensible approach, and I think it will fix a number of issues:</p>

<p><a href="https://issues.apache.org/jira/browse/CURATOR-448" title="Include curator framework state in error messages" class="issue-link" data-issue-key="CURATOR-448"><del>CURATOR-448</del></a>, <a href="https://issues.apache.org/jira/browse/CURATOR-467" title="Avoid logging errors after curator framework is closed." class="issue-link" data-issue-key="CURATOR-467">CURATOR-467</a> and <a href="https://issues.apache.org/jira/browse/CURATOR-468" title="Closing an active leader selector always logs spurious error" class="issue-link" data-issue-key="CURATOR-468"><del>CURATOR-468</del></a></p>

<p>Would you like to submit a PR for this fix?</p>
